### DISCUSSION TITLE
Can't use discriminatedUnion after `transform`

### DISCUSSION THREAD

**Discussion by alexgleason** [colinhacks/zod#2100] [Q&A] [answered]
I'm building a federated social media app (think like Mastodon) which validates external json data in [ActivityStreams](https://www.w3.org/TR/activitystreams-core/) format.

There are various competing ways to implement a quote post, for example. I want to consolidate them into one:

```ts
const noteSchema = z.object({
  type: z.literal('Note'),
  id: apId,
  to: recipients,
  cc: recipients,
  content: z.string(),
  attachment: z.array(attachmentSchema).optional().catch(undefined),
  inReplyTo: apId.optional(),
  attributedTo: apId,
  published: published(),

  // Quote posting
  quoteUrl: apId.optional(),
  quoteURL: apId.optional(),
  quoteUri: apId.optional(),
  _misskey_quote: apId.optional(),
}).transform((note) => {
  const { quoteUrl, quoteUri, quoteURL, _misskey_quote, ...rest } = note;

  return {
    // Consolidate the quote into one field
    quoteUrl: quoteUrl || quoteUri || quoteURL || _misskey_quote,
    ...rest,
  }
});
```

This mostly works, but since `transform` returns a `ZodEffect`, I can no longer do a discriminated union over it.

![image](https://user-images.githubusercontent.com/3639540/221367199-e2df53aa-8355-415d-801d-eeb8bbed924f.png)

![image](https://user-images.githubusercontent.com/3639540/221367215-68965113-80a4-4782-9973-b1f269d4efa6.png)

I've tried accessing `noteSchema._def.schema` for the discriminated union, as suggested in https://github.com/colinhacks/zod/issues/2056#issuecomment-1435690133, but this returns the _input schema_.

Is there a better way to do what I'm trying to do with Zod?

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const noteSchema = z.object( {
    type: z.literal( 'Note' ),
    id: apId,
    to: recipients,
    cc: recipients,
    content: z.string(),
    attachment: z.array( attachmentSchema ).optional().catch( undefined ),
    inReplyTo: apId.optional(),
    attributedTo: apId,
    published: published(),
    quoteUrl: apId.optional(),
} )

const flexibleNoteSchema = noteSchema.extend( {
    quoteURL: apId.optional(),
    quoteUri: apId.optional(),
    _misskey_quote: apId.optional(),
} ).transform( ( note ) => {
    const { quoteUrl, quoteUri, quoteURL, _misskey_quote, ...rest } = note
    return {
        quoteUrl: quoteUrl || quoteUri || quoteURL || _misskey_quote,
        ...rest,
    }
} )

const objectSchema = z.union( [
    flexibleNoteSchema,
    actorSchema,
] ).pipe(
    z.discriminatedUnion( 'type', [
        noteSchema,
        actorSchema,
    ] )
)
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by alexgleason** [reply]
Getting closer. Now `z.infer<typeof objectSchema>` only includes fields shared between actorSchema and noteSchema. Thanks for introducing me to `pipe()`, thinking about it :gear:

**Reply by alexgleason** [reply]
Actually my text editor was just freaking out. This seems to be working!

**Reply by JacobWeisenburger** [reply]
Thanks so much for the sponsorship! 🙏

**Comment by codingedgar** [comment]
For anyone else finding this issue, this is a limitation of zod at least until v4 https://github.com/colinhacks/zod/pull/2441#issuecomment-2044094604 

The accepted answer is a valid workaround, but I have the hunch that it throws the advantage of using `z.discriminatedUnion,` which is fast parsing.

An alternative, that might be more efficient could be

```ts
const noteSchema = z.object( {
    type: z.literal( 'Note' ),
    id: apId,
    to: recipients,
    cc: recipients,
    content: z.string(),
    attachment: z.array( attachmentSchema ).optional().catch( undefined ),
    inReplyTo: apId.optional(),
    attributedTo: apId,
    published: published(),
    quoteUrl: apId.optional(),
} );

const objectSchema = z.discriminatedUnion( 'type', [
        noteSchema,
        actorSchema,
    ])
    .transform((val) =>
         val.type === 'Note'
               // put the transforms after the switch
               ? ({...val, quoteUrl: quoteUrl || quoteUri || quoteURL || _misskey_quote})
               : val
      )
```

(Or just use `z.union` no point piping? not sure if zod is smart enough to avoid what seems to be a double parsing, even tho the types make sense there is a runtime cost involved)
